awscli/customizations/commands.py

```diff
@@ -134,7 +134,7 @@ def __call__(self, args, parsed_globals):
         event = 'before-building-argument-table-parser.%s' % \
             ".".join(self.lineage_names)
         self._session.emit(event, argument_table=self._arg_table, args=args,
-                           session=self._session)
```

Comment:
I know we had originally agreed upon `building-command-table`, but upon diving into the code, this seemed by far to be the less invasive approach. I do see now that it runs twice, once for the service and then for the operation, which is definitely behavior that should be avoided even if the handler is idempotent.

The trouble is figuring out a clean implementation of passing `parsed_globals` to that event. The `ServiceOperation` object only gets that variable when its invoked in a `ServiceCommand` object's `__call__` method, which is after `building-command-table` has run. I think its definitely doable, but would require a larger refactor, which I wanted to avoid for something that is pretty core behavior to the CLI.

The easiest way that is occurring to me now would be to pass `parsed_globals` to `_get_command_table`, which would pass it in turn to `_create_command_table`. Let me know what you think of that approach and if you have any concerns with it. Happy to investigate further, but if you do come up with any suggestions let me know.

Reply:
I think originally I had two different potential events in mind:

* `top-level-args-parsed` - This is the event for when the global args get parsed. It seemed suiting as we would just immediately set the global argument of `--endpoint-url` at that stage and we can keep endpoint url mutations isolated to that stage. The only issue with that approach is that the session had technically not been initialized yet, so there may be unintended consequences/configuration resolution if we try to grab the region before that. This event also happens before even logging is enabled so that is a concern as well if we are trying to do any tracing of it.
* `session-initialized` - This seemed like the next best event to use as the session was initialized there. The only problem I saw in using this event is that we could not 100% capture correctly the service being used. Specifically, for [CLI aliases](https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-alias.html) the command that gets parsed is represented as the alias and not the underlying service command at that point. So we would be using the wrong endpoint if the command was under an alias.

This all being said. I don't know if I have a clear preference between `building-command-table` and `before-building-argument-table-parser`. They both share similar concerns, but it seems like we need an event lower into the stack to get this to work appropriately and I do not want to be adding a new event for this. I think for now let's just stick with `before-building-argument-table-parser` unless we get a better idea.

Reply:
Ok cool sounds good Kyle. I'll keep the behavior as is. I'll be adding the additional services that were missed in my script that you caught in yours shortly.